File: app/local_plugins/service_manager/android/src/main/kotlin/com/aiagent/service_manager/ServiceManagerPlugin.kt
package com.aiagent.service_manager

import android.content.Context
import android.util.Log
import com.aiagent.plugin_interface.ServiceTestEventSink
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.plugin.common.EventChannel
import io.flutter.plugin.common.MethodChannel
import kotlinx.coroutines.*

/**
 * ServiceManagerPlugin — MethodChannel/EventChannel 调度层
 *
 * 职责：
 * 1. 管理 MethodChannel（Flutter → Native 命令路由）
 * 2. 管理 EventChannel（Native → Flutter 事件转发）
 * 3. 创建 ServiceTestRunner，将命令委托过去
 *
 * 对应 AgentsServerPlugin 的角色，但管的是服务测试而不是 Agent。
 */
class ServiceManagerPlugin : FlutterPlugin, ServiceTestEventSink {

    companion object {
        private const val TAG = "ServiceManagerPlugin"
    }

    private lateinit var methodChannel: MethodChannel
    private lateinit var eventChannel: EventChannel
    private var eventSinkStream: EventChannel.EventSink? = null

    private val mainScope = CoroutineScope(SupervisorJob() + Dispatchers.Main)
    private lateinit var context: Context
    private var runner: ServiceTestRunner? = null

    // ─────────────────────────────────────────────────
    // FlutterPlugin 生命周期
    // ─────────────────────────────────────────────────

    override fun onAttachedToEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        context = binding.applicationContext
        runner = ServiceTestRunner(context, this)

        methodChannel = MethodChannel(binding.binaryMessenger, "service_manager/commands")
        methodChannel.setMethodCallHandler { call, result ->
            val r = runner
            if (r == null) {
                Log.w(TAG, "Runner not ready, ignoring ${call.method}")
                result.success(null)
                return@setMethodCallHandler
            }

            when (call.method) {
                // ── STT ──
                "testSttStart" -> {
                    r.testSttStart(
                        testId = call.argument<String>("testId")!!,
                        serviceId = call.argument<String>("serviceId")!!,
                    )
                    result.success(null)
                }
                "testSttStop" -> {
                    r.testSttStop(call.argument<String>("testId")!!)
                    result.success(null)
                }

                // ── TTS ──
                "testTtsSpeak" -> {
                    r.testTtsSpeak(
                        testId = call.argument<String>("testId")!!,
                        serviceId = call.argument<String>("serviceId")!!,
                        text = call.argument<String>("text")!!,
                        voiceName = call.argument<String>("voiceName"),
                        speed = call.argument<Double>("speed") ?: 1.0,
                        pitch = call.argument<Double>("pitch") ?: 1.0,
                    )
                    result.success(null)
                }
                "testTtsStop" -> {
                    r.testTtsStop(call.argument<String>("testId")!!)
                    result.success(null)
                }

                // ── LLM ──
                "testLlmChat" -> {
                    r.testLlmChat(
                        testId = call.argument<String>("testId")!!,
                        serviceId = call.argument<String>("serviceId")!!,
                        text = call.argument<String>("text")!!,
                    )
                    result.success(null)
                }
                "testLlmCancel" -> {
                    r.testLlmCancel(call.argument<String>("testId")!!)
                    result.success(null)
                }

                // ── Translation ──
                "testTranslate" -> {
                    r.testTranslate(
                        testId = call.argument<String>("testId")!!,
                        serviceId = call.argument<String>("serviceId")!!,
                        text = call.argument<String>("text")!!,
                        targetLang = call.argument<String>("targetLang")!!,
                        sourceLang = call.argument<String>("sourceLang"),
                    )
                    result.success(null)
                }

                // ── STS ──
                "testStsConnect" -> {
                    r.testStsConnect(
                        testId = call.argument<String>("testId")!!,
                        serviceId = call.argument<String>("serviceId")!!,
                    )
                    result.success(null)
                }
                "testStsStartAudio" -> {
                    r.testStsStartAudio(call.argument<String>("testId")!!)
                    result.success(null)
                }
                "testStsStopAudio" -> {
                    r.testStsStopAudio(call.argument<String>("testId")!!)
                    result.success(null)
                }
                "testStsDisconnect" -> {
                    r.testStsDisconnect(call.argument<String>("testId")!!)
                    result.success(null)
                }

                // ── AST ──
                "testAstConnect" -> {
                    r.testAstConnect(
                        testId = call.argument<String>("testId")!!,
                        serviceId = call.argument<String>("serviceId")!!,
                    )
                    result.success(null)
                }
                "testAstStartAudio" -> {
                    r.testAstStartAudio(call.argument<String>("testId")!!)
                    result.success(null)
                }
                "testAstStopAudio" -> {
                    r.testAstStopAudio(call.argument<String>("testId")!!)
                    result.success(null)
                }
                "testAstDisconnect" -> {
                    r.testAstDisconnect(call.argument<String>("testId")!!)
                    result.success(null)
                }

                // ── 自动化测试 ──
                "autoTest" -> {
                    r.autoTest(
                        testId = call.argument<String>("testId")!!,
                        serviceId = call.argument<String>("serviceId")!!,
                    )
                    result.success(null)
                }

                // ── 通用 ──
                "releaseTest" -> {
                    r.releaseSession(call.argument<String>("testId")!!)
                    result.success(null)
                }

                else -> result.notImplemented()
            }
        }

        eventChannel = EventChannel(binding.binaryMessenger, "service_manager/events")
        eventChannel.setStreamHandler(object : EventChannel.StreamHandler {
            override fun onListen(arguments: Any?, events: EventChannel.EventSink?) {
                eventSinkStream = events
            }
            override fun onCancel(arguments: Any?) {
                eventSinkStream = null
            }
        })
    }

    override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        methodChannel.setMethodCallHandler(null)
        eventChannel.setStreamHandler(null)
        runner?.releaseAll()
        runner = null
        mainScope.cancel()
    }

    // ─────────────────────────────────────────────────
    // ServiceTestEventSink 实现（Runner → Plugin → EventChannel → Flutter）
    // ─────────────────────────────────────────────────

    override fun onSttTestEvent(testId: String, kind: String, text: String?, errorCode: String?, errorMessage: String?) {
        pushEvent(mapOf(
            "type" to "stt",
            "testId" to testId,
            "kind" to kind,
            "text" to text,
            "errorCode" to errorCode,
            "errorMessage" to errorMessage,
        ))
    }

    override fun onTtsTestEvent(testId: String, kind: String, progressMs: Int?, durationMs: Int?, errorCode: String?, errorMessage: String?) {
        pushEvent(mapOf(
            "type" to "tts",
            "testId" to testId,
            "kind" to kind,
            "progressMs" to progressMs,
            "durationMs" to durationMs,
            "errorCode" to errorCode,
            "errorMessage" to errorMessage,
        ))
    }

    override fun onLlmTestEvent(testId: String, kind: String, textDelta: String?, thinkingDelta: String?, toolCallId: String?, toolName: String?, toolArgumentsDelta: String?, toolResult: String?, fullText: String?, errorCode: String?, errorMessage: String?) {
        pushEvent(mapOf(
            "type" to "llm",
            "testId" to testId,
            "kind" to kind,
            "textDelta" to textDelta,
            "thinkingDelta" to thinkingDelta,
            "toolCallId" to toolCallId,
            "toolName" to toolName,
            "toolArgumentsDelta" to toolArgumentsDelta,
            "toolResult" to toolResult,
            "fullText" to fullText,
            "errorCode" to errorCode,
            "errorMessage" to errorMessage,
        ))
    }

    override fun onTranslationTestEvent(testId: String, kind: String, sourceText: String?, translatedText: String?, sourceLanguage: String?, targetLanguage: String?, errorCode: String?, errorMessage: String?) {
        pushEvent(mapOf(
            "type" to "translation",
            "testId" to testId,
            "kind" to kind,
            "sourceText" to sourceText,
            "translatedText" to translatedText,
            "sourceLanguage" to sourceLanguage,
            "targetLanguage" to targetLanguage,
            "errorCode" to errorCode,
            "errorMessage" to errorMessage,
        ))
    }

    override fun onStsTestEvent(testId: String, kind: String, text: String?, state: String?, errorCode: String?, errorMessage: String?) {
        pushEvent(mapOf(
            "type" to "sts",
            "testId" to testId,
            "kind" to kind,
            "text" to text,
            "state" to state,
            "errorCode" to errorCode,
            "errorMessage" to errorMessage,
        ))
    }

    override fun onAstTestEvent(testId: String, kind: String, text: String?, state: String?, errorCode: String?, errorMessage: String?) {
        pushEvent(mapOf(
            "type" to "ast",
            "testId" to testId,
            "kind" to kind,
            "text" to text,
            "state" to state,
            "errorCode" to errorCode,
            "errorMessage" to errorMessage,
        ))
    }

    override fun onTestDone(testId: String, success: Boolean, message: String?) {
        pushEvent(mapOf(
            "type" to "done",
            "testId" to testId,
            "success" to success,
            "message" to message,
        ))
    }

    private fun pushEvent(data: Map<String, Any?>) {
        mainScope.launch { eventSinkStream?.success(data) }
    }
}

File: apps/client/local_plugins/agents/service_manager/ios/Classes/ServiceManagerPlugin.swift
import Flutter
import ai_plugin_interface
import os.log

/// iOS counterpart of the Android `ServiceManagerPlugin`.
///
/// Routes Dart commands on `service_manager/commands` to a `ServiceTestRunner`,
/// and fans `ServiceTestEventSink` callbacks back through
/// `service_manager/events`. The event payload schema mirrors the Android
/// side so the Dart bridge stays platform-agnostic.
public final class ServiceManagerPlugin: NSObject, FlutterPlugin, ServiceTestEventSink {
    private static let log = OSLog(subsystem: "com.aiagent.service_manager", category: "Plugin")

    private var methodChannel: FlutterMethodChannel?
    private var eventChannel: FlutterEventChannel?
    private let streamHandler = SmStreamHandler()
    private var runner: ServiceTestRunner!

    public static func register(with registrar: FlutterPluginRegistrar) {
        let instance = ServiceManagerPlugin()
        instance.runner = ServiceTestRunner(eventSink: instance)
        let messenger = registrar.messenger()

        instance.methodChannel = FlutterMethodChannel(
            name: "service_manager/commands",
            binaryMessenger: messenger
        )
        registrar.addMethodCallDelegate(instance, channel: instance.methodChannel!)

        instance.eventChannel = FlutterEventChannel(
            name: "service_manager/events",
            binaryMessenger: messenger
        )
        instance.eventChannel?.setStreamHandler(instance.streamHandler)
    }

    public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        let args = (call.arguments as? [String: Any?]) ?? [:]
        switch call.method {

        // ── LLM ──
        case "testLlmChat":
            guard let testId = args["testId"] as? String,
                  let serviceId = args["serviceId"] as? String,
                  let text = args["text"] as? String else {
                result(FlutterError(code: "INVALID_ARGUMENT", message: "testLlmChat args", details: nil)); return
            }
            runner.testLlmChat(testId: testId, serviceId: serviceId, text: text)
            result(nil)
        case "testLlmCancel":
            guard let testId = args["testId"] as? String else {
                result(FlutterError(code: "INVALID_ARGUMENT", message: "testId missing", details: nil)); return
            }
            runner.testLlmCancel(testId: testId)
            result(nil)

        // ── Translation ──
        case "testTranslate":
            guard let testId = args["testId"] as? String,
                  let serviceId = args["serviceId"] as? String,
                  let text = args["text"] as? String,
                  let targetLang = args["targetLang"] as? String else {
                result(FlutterError(code: "INVALID_ARGUMENT", message: "testTranslate args", details: nil)); return
            }
            let sourceLang = args["sourceLang"] as? String
            runner.testTranslate(testId: testId, serviceId: serviceId, text: text,
                                 targetLang: targetLang, sourceLang: sourceLang)
            result(nil)

        // ── Not wired on iOS yet: surface a structured error rather than
        //    silently succeeding so the UI can mark the test as failed. ──
        case "testSttStart":
            guard let testId = args["testId"] as? String else { result(nil); return }
            runner.notImplemented(testId: testId, area: "stt"); result(nil)
        case "testTtsSpeak":
            guard let testId = args["testId"] as? String else { result(nil); return }
            runner.notImplemented(testId: testId, area: "tts"); result(nil)
        case "testStsConnect":
            guard let testId = args["testId"] as? String else { result(nil); return }
            runner.notImplemented(testId: testId, area: "sts"); result(nil)
        case "testAstConnect":
            guard let testId = args["testId"] as? String else { result(nil); return }
            runner.notImplemented(testId: testId, area: "ast"); result(nil)

        // Stop/cancel methods are no-ops while the underlying paths are stubs.
        case "testSttStop", "testTtsStop",
             "testStsStartAudio", "testStsStopAudio", "testStsDisconnect",
             "testAstStartAudio", "testAstStopAudio", "testAstDisconnect":
            result(nil)

        default:
            result(FlutterMethodNotImplemented)
        }
    }

    // ── ServiceTestEventSink ──────────────────────────────────────

    public func onSttTestEvent(testId: String, kind: String, text: String?, errorCode: String?, errorMessage: String?) {
        push([
            "type": "stt",
            "testId": testId,
            "kind": kind,
            "text": text as Any,
            "errorCode": errorCode as Any,
            "errorMessage": errorMessage as Any,
        ])
    }

    public func onTtsTestEvent(testId: String, kind: String, progressMs: Int?, durationMs: Int?, errorCode: String?, errorMessage: String?) {
        push([
            "type": "tts",
            "testId": testId,
            "kind": kind,
            "progressMs": progressMs as Any,
            "durationMs": durationMs as Any,
            "errorCode": errorCode as Any,
            "errorMessage": errorMessage as Any,
        ])
    }

    public func onLlmTestEvent(
        testId: String,
        kind: String,
        textDelta: String?,
        thinkingDelta: String?,
        toolCallId: String?,
        toolName: String?,
        toolArgumentsDelta: String?,
        toolResult: String?,
        fullText: String?,
        errorCode: String?,
        errorMessage: String?
    ) {
        push([
            "type": "llm",
            "testId": testId,
            "kind": kind,
            "textDelta": textDelta as Any,
            "thinkingDelta": thinkingDelta as Any,
            "toolCallId": toolCallId as Any,
            "toolName": toolName as Any,
            "toolArgumentsDelta": toolArgumentsDelta as Any,
            "toolResult": toolResult as Any,
            "fullText": fullText as Any,
            "errorCode": errorCode as Any,
            "errorMessage": errorMessage as Any,
        ])
    }

    public func onTranslationTestEvent(
        testId: String,
        kind: String,
        sourceText: String?,
        translatedText: String?,
        sourceLanguage: String?,
        targetLanguage: String?,
        errorCode: String?,
        errorMessage: String?
    ) {
        push([
            "type": "translation",
            "testId": testId,
            "kind": kind,
            "sourceText": sourceText as Any,
            "translatedText": translatedText as Any,
            "sourceLanguage": sourceLanguage as Any,
            "targetLanguage": targetLanguage as Any,
            "errorCode": errorCode as Any,
            "errorMessage": errorMessage as Any,
        ])
    }

    public func onStsTestEvent(testId: String, kind: String, text: String?, state: String?, errorCode: String?, errorMessage: String?) {
        push([
            "type": "sts", "testId": testId, "kind": kind,
            "text": text as Any, "state": state as Any,
            "errorCode": errorCode as Any, "errorMessage": errorMessage as Any,
        ])
    }

    public func onAstTestEvent(testId: String, kind: String, text: String?, state: String?, errorCode: String?, errorMessage: String?) {
        push([
            "type": "ast", "testId": testId, "kind": kind,
            "text": text as Any, "state": state as Any,
            "errorCode": errorCode as Any, "errorMessage": errorMessage as Any,
        ])
    }

    public func onTestDone(testId: String, success: Bool, message: String?) {
        push([
            "type": "done",
            "testId": testId,
            "success": success,
            "message": message as Any,
        ])
    }

    private func push(_ payload: [String: Any]) {
        streamHandler.send(payload)
    }
}

/// Local copy of the stream handler — keeps service_manager independent of
/// the agents_server module.
private final class SmStreamHandler: NSObject, FlutterStreamHandler {
    private let lock = NSLock()
    private var sink: FlutterEventSink?

    func send(_ data: [String: Any]) {
        lock.lock(); let s = sink; lock.unlock()
        guard let s = s else { return }
        if Thread.isMainThread { s(data) }
        else { DispatchQueue.main.async { s(data) } }
    }

    func onListen(withArguments arguments: Any?, eventSink events: @escaping FlutterEventSink) -> FlutterError? {
        lock.lock(); sink = events; lock.unlock(); return nil
    }

    func onCancel(withArguments arguments: Any?) -> FlutterError? {
        lock.lock(); sink = nil; lock.unlock(); return nil
    }
}
